ext/bcmath/bcmath.c

```diff
@@ -777,4 +791,943 @@ PHP_FUNCTION(bcscale)
+{
+	if (check_empty == ZEND_PROPERTY_NOT_EMPTY) {
+		bcmath_number_obj_t *intern = get_bcmath_number_from_obj(obj);
+
+		if (zend_string_equals_literal(name, "value")) {
+			return !bc_is_zero(intern->num);
+		}
+
+		if (zend_string_equals_literal(name, "scale")) {
+			return intern->scale != 0;
+		}
+	}
+	return zend_string_equals_literal(name, "value") || zend_string_equals_literal(name, "scale");
+}
+
+static zend_result bcmath_number_cast_object(zend_object *obj, zval *ret, int type)
+{
+	if (type == _IS_BOOL) {
+		bcmath_number_obj_t *intern = get_bcmath_number_from_obj(obj);
+		ZVAL_BOOL(ret, !bc_is_zero(intern->num));
+		return SUCCESS;
+	}
+
+	return zend_std_cast_object_tostring(obj, ret, type);
+}
+
+static void bcmath_number_register_class(void)
+{
+	bcmath_number_ce = register_class_BcMath_Number(zend_ce_stringable);
+	bcmath_number_ce->create_object = bcmath_number_create;
+	bcmath_number_ce->default_object_handlers = &bcmath_number_obj_handlers;
+
+	memcpy(&bcmath_number_obj_handlers, &std_object_handlers, sizeof(zend_object_handlers));
+	bcmath_number_obj_handlers.offset = XtOffsetOf(bcmath_number_obj_t, std);
+	bcmath_number_obj_handlers.free_obj = bcmath_number_free;
+	bcmath_number_obj_handlers.clone_obj = bcmath_number_clone;
+	bcmath_number_obj_handlers.do_operation = bcmath_number_do_operation;
+	bcmath_number_obj_handlers.compare = bcmath_number_compare;
+	bcmath_number_obj_handlers.write_property = bcmath_number_write_property;
+	bcmath_number_obj_handlers.unset_property = bcmath_number_unset_property;
+	bcmath_number_obj_handlers.has_property = bcmath_number_has_property;
+	bcmath_number_obj_handlers.read_property = bcmath_number_read_property;
+	bcmath_number_obj_handlers.get_properties_for = bcmath_number_get_properties_for;
+	bcmath_number_obj_handlers.cast_object = bcmath_number_cast_object;
+}
+
+static zend_always_inline void bcmath_number_add_internal(
+	bc_num n1, bc_num n2, bc_num *ret,
+	size_t n1_full_scale, size_t n2_full_scale, size_t *scale, bool auto_scale
+) {
+	if (auto_scale) {
+		*scale = MAX(n1_full_scale, n2_full_scale);
+	}
+	*ret = bc_add(n1, n2, *scale);
+	(*ret)->n_scale = MIN(*scale, (*ret)->n_scale);
+	bc_rm_trailing_zeros(*ret);
+}
+
+static zend_always_inline void bcmath_number_sub_internal(
+	bc_num n1, bc_num n2, bc_num *ret,
+	size_t n1_full_scale, size_t n2_full_scale, size_t *scale, bool auto_scale
+) {
+	if (auto_scale) {
+		*scale = MAX(n1_full_scale, n2_full_scale);
+	}
+	*ret = bc_sub(n1, n2, *scale);
+	(*ret)->n_scale = MIN(*scale, (*ret)->n_scale);
+	bc_rm_trailing_zeros(*ret);
+}
+
+static zend_always_inline zend_result bcmath_number_mul_internal(
+	bc_num n1, bc_num n2, bc_num *ret,
+	size_t n1_full_scale, size_t n2_full_scale, size_t *scale, bool auto_scale
+) {
+	if (auto_scale) {
+		*scale = n1_full_scale + n2_full_scale;
+		if (UNEXPECTED(CHECK_RET_SCALE_OVERFLOW(*scale, n1_full_scale))) {
```

Comment:
Unless I missed it, I don't see a test case for this, not sure if _strictly_ necessary though.

Reply:
I tried the test, but it turned out to be a pretty slow test.
This is a rare case that cannot be called common, so I will ignore it for now.